flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java

```diff
@@ -82,49 +94,107 @@
         }
     }
 
-    @Test
-    public void testNotRegisterMetrics() throws Exception {
+    @ParameterizedTest
+    @EnumSource(DeliveryGuarantee.class)
+    public void testNotRegisterMetrics(DeliveryGuarantee guarantee) throws Exception {
         final Properties config = getKafkaClientConfiguration();
         config.put("flink.disable-metrics", "true");
-        try (final KafkaWriter<Integer> ignored = createWriterWithConfiguration(config)) {
+        try (final KafkaWriter<Integer> ignored =
+                createWriterWithConfiguration(config, guarantee)) {
             assertThrows(
                     IllegalArgumentException.class,
                     () -> metricListener.getGauge(KAFKA_METRIC_WITH_GROUP_NAME));
         }
     }
 
-    private KafkaWriter<Integer> createWriterWithConfiguration(Properties config) {
+    @Test
+    public void testIncreasingByteOutCounter() throws Exception {
+        final OperatorIOMetricGroup operatorIOMetricGroup = new NumBytesOutOperatorIOMetricGroup();
+        final InternalSinkWriterMetricGroup metricGroup =
+                InternalSinkWriterMetricGroup.mock(
+                        metricListener.getMetricGroup(), operatorIOMetricGroup);
+        try (final KafkaWriter<Integer> writer =
+                createWriterWithConfiguration(
+                        getKafkaClientConfiguration(), DeliveryGuarantee.NONE, metricGroup)) {
+            final Counter numBytesOut = operatorIOMetricGroup.getNumBytesOutCounter();
+            Assertions.assertEquals(numBytesOut.getCount(), 0L);
+            writer.write(1, SINK_WRITER_CONTEXT);
+            metricListener.getGauge("currentSendTime");
+            MatcherAssert.assertThat(numBytesOut.getCount(), greaterThan(0L));
+        }
+    }
+
+    @Test
+    public void testCurrentSendTimeMetric() throws Exception {
+        final OperatorIOMetricGroup operatorIOMetricGroup = new NumBytesOutOperatorIOMetricGroup();
+        final InternalSinkWriterMetricGroup metricGroup =
+                InternalSinkWriterMetricGroup.mock(
+                        metricListener.getMetricGroup(), operatorIOMetricGroup);
+        try (final KafkaWriter<Integer> writer =
+                createWriterWithConfiguration(
+                        getKafkaClientConfiguration(),
+                        DeliveryGuarantee.AT_LEAST_ONCE,
+                        metricGroup)) {
+            final Gauge<Long> currentSendTime = metricListener.getGauge("currentSendTime");
+            Assertions.assertEquals(currentSendTime.getValue(), 0L);
+            IntStream.range(0, 100)
+                    .forEach(
+                            (run) -> {
+                                try {
+                                    writer.write(1, SINK_WRITER_CONTEXT);
+                                    // Manually flush the records to generate a sendTime
+                                    if (run % 10 == 0) {
+                                        writer.prepareCommit(false);
```

Comment:
How about calling a `VisibleForTesting` only flush method here?

Reply:
I am very hesitant to use `@VisibleForTesting` because it always feels to me like proof of bad design if something cannot be tested without it. In this case, I'd refrain from adding it. WDYT?

Reply:
I'm fine with leaving as is.